tower/src/util/clone_boxed.rs

```diff
@@ -0,0 +1,138 @@
+///         println!("received request");
+///         req
+///     })
+///     .map_response(|res| {
+///         println!("response produced");
+///         res
+///     })
+///     .load_shed()
+///     .concurrency_limit(64)
+///     .timeout(Duration::from_secs(10))
+///     .service_fn(|req: Request| async {
+///         Ok::<_, BoxError>(Response::new())
+///     });
+/// # let service = assert_service(service);
+///
+/// // `CloneService` will erase the type so its nameable
+/// let service: CloneBoxService<Request, Response, BoxError> = CloneBoxService::new(service);
+/// # let service = assert_service(service);
+///
+/// // And we can still clone the service
+/// let cloned_service = service.clone();
+/// #
+/// # fn assert_service<S, R>(svc: S) -> S
+/// # where S: Service<R> { svc }
+/// ```
+pub struct CloneBoxService<T, U, E>(
+    Box<
+        dyn CloneService<T, Response = U, Error = E, Future = BoxFuture<'static, Result<U, E>>>
+            + Send,
+    >,
+);
+
+impl<T, U, E> CloneBoxService<T, U, E> {
+    /// Create a new `CloneBoxService`.
+    pub fn new<S>(inner: S) -> Self
+    where
+        S: Service<T, Response = U, Error = E> + Clone + Send + 'static,
+        S::Future: Send + 'static,
+    {
+        let inner = inner.map_future(|f| Box::pin(f) as _);
+        CloneBoxService(Box::new(inner))
+    }
+
+    /// Returns a [`Layer`] for wrapping a [`Service`] in a [`CloneBoxService`]
+    /// middleware.
+    ///
+    /// [`Layer`]: crate::Layer
+    pub fn layer<S>() -> LayerFn<fn(S) -> Self>
+    where
+        S: Service<T, Response = U, Error = E> + Clone + Send + 'static,
+        S::Future: Send + 'static,
+    {
+        layer_fn(Self::new)
+    }
+}
+
+impl<T, U, E> Service<T> for CloneBoxService<T, U, E> {
+    type Response = U;
+    type Error = E;
+    type Future = BoxFuture<'static, Result<U, E>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), E>> {
+        self.0.poll_ready(cx)
+    }
+
+    #[inline]
+    fn call(&mut self, request: T) -> Self::Future {
+        self.0.call(request)
+    }
+}
+
+impl<T, U, E> Clone for CloneBoxService<T, U, E> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone_box())
+    }
+}
+
+trait CloneService<R>: Service<R> {
+    fn clone_box(
+        &self,
+    ) -> Box<
+        dyn CloneService<R, Response = Self::Response, Error = Self::Error, Future = Self::Future>
+            + Send,
+    >;
+}
+
+impl<R, T> CloneService<R> for T
+where
+    T: Service<R> + Send + Clone + 'static,
+{
+    fn clone_box(
+        &self,
+    ) -> Box<dyn CloneService<R, Response = T::Response, Error = T::Error, Future = T::Future> + Send>
+    {
+        Box::new(self.clone())
+    }
+}
+
+impl<T, U, E> fmt::Debug for CloneBoxService<T, U, E>
+where
+    T: fmt::Debug,
+    U: fmt::Debug,
+    E: fmt::Debug,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("CloneBoxService").finish()
```

Comment:
take it or leave it: we _may_ want to consider including the request and response type _names_ here, so different instances of `CloneBoxService` can be distinguished? Maybe something like
```suggestion
        write!(
            fmt, 
            "CloneBoxService<Request = {}, Response = {}, Error = {}>",
            std::any::type_name::<T>(),
            std::any::type_name::<U>(),
            std::any::type_name::<E>(),
        )
```
but, on the other hand, this could be quite long, and we might want to correctly handle multi-line formatting in alt mode (`{:#}`)...so, maybe that's something to hold off on and add later if it's needed...

Reply:
also, `BoxService` doesn't do this, so it's probably better left for a separate PR if we _do_ want to add something like that...

Reply:
I was actually thinking the type params are there such that this could be improved in the future, without a breaking change. I think I'll leave it as it is now and make an issue about improving the formatting, for `BoxService` as well.

Reply:
> I was actually thinking the type params are there such that this could be improved in the future,

I don't think we want the type params to be `Debug` in any case; we can't actually include values of those types in the formatted output, since they're not part of the struct. I think the where clause won't ever be useful for future improvements to the debug format, since we'll never have instances of those types to format when formatting a `CloneBoxService` instance. We can use `type_name` to get the _names_ of those types without requiring those types to be `Debug`.

Reply:
Oh yeah doh of course! I've removed the bounds.